# Exercise 2

In this exercise, you will complete the implementation of a Decision Tree classifier based on our simple `fduml` framework. We have written most of the code for you already, and you only need to fill in the most essential parts marked in `TODO`. We have also prepared several test cases for you to check if your code works correctly. Furthermore, you can also test the accuracy of your code by comparing its output with the output of Sklearn.

In [1]:
# Auto reload external modules, which means you can modify the code of our fduml implementation without restarting the kernel.
%load_ext autoreload
%autoreload 2

In [2]:
# Basic imports
import numpy as np
import random
import matplotlib.pyplot as plt
from matplotlib import rc

np.random.seed(42)
random.seed(42)

## Implement and test (40 points)

We have prepared several test cases for you to check if your code works correctly. After you write your own implementation, try the following code for testing.

In [3]:
from fduml import *

In [4]:
from fduml.tree.tests.test_decision_tree import test_dt_classification
test_dt_classification()

[[-2 -1]
 [-1 -1]
 [-1 -2]
 [ 1  1]
 [ 1  2]
 [ 2  1]]


## Load data and fit the model (40 points)

Inside the `data` directory we have prepared a classification dataset, split into training and test sets. In this part, you will load the data and fit the model to the training data. Then, you will evaluate the model on the test data.

In [5]:
# Load the water potability dataset
import pandas as pd

train_data = pd.read_csv('../Exercise2/data/water_potability_train.csv')
test_data = pd.read_csv('../Exercise2/data/water_potability_test.csv')

print("train_data\n")
print(train_data)

print("test_data\n")
print(test_data)

train_data

            ph    Hardness       Solids  Chloramines     Sulfate  \
0     8.618654  257.595883  11595.35498     6.399933   -1.000000   
1    -1.000000  199.942222  25973.32663     6.490994  336.040741   
2     4.923179  208.406673  15990.14923     5.648146  349.655175   
3     7.617524  179.596189  30308.23118     6.952617  329.422414   
4     8.891674  184.869606  41801.44184     3.409576  337.047108   
...        ...         ...          ...          ...         ...   
2616  4.814136  205.214041  17650.40505     8.121080  350.487939   
2617 -1.000000  221.391974  27979.73622     6.572390  349.624553   
2618  5.596730  229.295098  44652.36387     6.500953  323.999049   
2619  7.079304  137.007355  24282.15477     5.705693  433.633900   
2620 -1.000000  255.953599  15097.02406     8.482421  361.971419   

      Conductivity  Organic_carbon  Trihalomethanes  Turbidity  Potability  
0       343.740007       15.331166        75.687401   4.141342           0  
1       344.97036

In [6]:
# Fit a DecisionTreeClassifier to the water potability train set
from fduml.tree import DecisionTreeClassifier
from time import time

X_train = train_data.drop(columns=['Potability'])
y_train = train_data['Potability']

X_test = test_data.drop(columns=['Potability'])
y_test = test_data['Potability']

start_time = time()
dt_classfier =  DecisionTreeClassifier(criterion="info_gain", random_state=0)
dt_classfier.fit(X_train.values, y_train.values)

y_pred = dt_classfier.predict(X_test.values)
end_time = time()

In [7]:
# Evaluate the DecisionTreeClassifier on the water potability test set
accuracy = accuracy_score(y_test.values, y_pred)
print(f'Accuracy of the Decision Tree Classifier: {accuracy:.2f}')
print(f'Classification took {end_time - start_time:.2f} seconds')

Accuracy of the Decision Tree Classifier: 0.59
Classification took 157.98 seconds


## Compare with Sklearn (20 points)

Since the interface of our `fduml` is the same as that of sklearn, you can easily compare the output of your implementation with that of sklearn. In this part, try to generate test data and compare the accuracy and running time of your implementation with that of sklearn. You can use the following code for comparison.

In the conclusion part, try to answer the following questions:

- Is the accuracy of your implementation the same as that of sklearn? If not, what can be the reason?

- Is the running time of your implementation the same as that of sklearn? If not, what can be the reason?

- If there is any special thing you want to mention, please write it down.

Note that we do not require you to match the accuracy and running time of sklearn (which can be quite difficult), but you should be able to explain the reason if they are different.

In [8]:
# Import necessary libraries
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from time import time


# Create and fit the Decision Tree classifier
start_time = time()
dt_classifier = DecisionTreeClassifier(criterion='gini', random_state=0)
dt_classifier.fit(X_train, y_train)

# Make predictions on the test data
y_pred = dt_classifier.predict(X_test)
end_time = time()

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)

# Print the accuracy
print(f'Accuracy of the Decision Tree classifier: {accuracy:.2f}')
print(f'Classification took {end_time - start_time:.2f} seconds')


Accuracy of the Decision Tree classifier: 0.61
Classification took 0.03 seconds


## Conclusion

#### Is the accuracy of your implementation the same as that of sklearn? If not, what can be the reason?



  Yes, the accuracy of my implementation is basically consistent with that of sklearn on the same test set. My accracy is 0.59 and the accuracy of sklearn is 0.61.
<br><br>



#### Is the running time of your implementation the same as that of sklearn? If not, what can be the reason?



  No, my implementation took 157.98 seconds while sklearn only took 0.03 seconds. Sklearn is 5266 times faster than mine :-)<br>
  Here are the reasons why scikit-learn is so fast:

- Cython and C/C++ Integration:
Cython: Many parts of scikit-learn are written in Cython, which is a superset of Python that allows you to write C-like performance-critical code while retaining Python’s syntax. This makes functions execute faster than pure Python.
C/C++ Integration: The most computationally intensive parts, like matrix operations and numerical computations, are often implemented using low-level C or C++ libraries. This allows scikit-learn to take advantage of the speed of these lower-level languages.<br><br>
- Use of Optimized Libraries:
scikit-learn leverages highly optimized numerical libraries like NumPy and SciPy, which themselves rely on low-level libraries such as BLAS (Basic Linear Algebra Subprograms) and LAPACK. These libraries are highly optimized for matrix operations and numerical computations.
By using these libraries, scikit-learn takes advantage of parallelized and optimized algorithms for common operations like matrix multiplication, solving systems of linear equations, etc.<br><br>
- Efficient Algorithms:
scikit-learn implements highly efficient algorithms, including those for decision trees, SVMs, and ensemble methods like random forests. The algorithms are well-researched, and they are implemented with a focus on efficiency.
The library also uses optimizations like data structures tailored for fast access (e.g., KD-trees for nearest neighbor searches).<br><br>
- Parallel Processing:
For many of its models, scikit-learn can use parallel processing to leverage multiple CPU cores, which can significantly speed up training and prediction, especially with ensemble methods like random forests or gradient boosting.<br><br>
- Compiled Code Execution:
By relying on compiled code through Cython or external libraries, scikit-learn avoids the Global Interpreter Lock (GIL) bottleneck of Python. This allows it to execute more efficiently, especially for tasks that require heavy computation.
This combination of optimized low-level code, efficient algorithms, and parallelism makes scikit-learn a powerful and fast tool for machine learning, even though its API is written in Python.